### ISSUE TITLE
Applications are not properly detected when used from within certain types of profiles.

### ISSUE THREAD

**Issue by pressRtowin** [TrianguloY/UrlChecker#184]
### Steps to reproduce

This is a bit of a weird one and I suggest taking a look at these two issues for more context:

- https://github.com/butzist/ActivityLauncher/issues/193
- https://github.com/tasomaniac/OpenLinkWith/issues/95

I'm not a developer, so frankly, _I'm not sure what's causing the behavior I'm about to describe_ but I've tested it fairly extensively so I'll provide as much information as I can here. 

But essentially, I use apps like OpenLinkWith—and since discovering it, URL Checker—for an additional purpose outside their standard use cases. On most flagship Samsung phones and tablets, you're able to have up to 4 copies of the same application installed (though it takes a bit of trickery to get the 4th) by using the main user account, a Work profile, Secure Folder, and Samsung's Dual Messenger feature.

By having a copy of URL Checker installed in each profile, I'm able to send links and open apps between profiles much more easily than usually available. The method varies a bit depending on the profile. It's easy to go between the Secure Folder and Main profiles for instance, as Samsung includes an item in the Share and Open With menus in both profiles that allows access to the other's apps. The work profile is a bit trickier, as Island's work profile has security settings in place that block most "Open with" actions from accessing Work apps, but as "Share" actions don't have this limitation, I can simply use [Share & Open](https://play.google.com/store/apps/details?id=ru.gelin.android.shareopen&hl=en_AU&gl=US&pli=1) as a workaround.

Other than some security differences as mentioned, these two profiles work very much as you'd expect any managed profile on Android to behave. But where things get really weird is the Dual Messenger profile. Firstly, by default, the feature is limited to only allowing certain social media apps to be installed, but it's easy to work around this issue by using ADB to install a package installer like [SAI](https://play.google.com/store/apps/details?id=com.aefyr.sai&hl=en_AU&gl=US) to the Dual Messenger profile, and then using it going forward to install whatever you want.

Once you're able to install whatever you want, you'll notice that some apps just behave a bit _weird_. These apps very much behave as if the barrier between profiles has been lessened in some way. Unlike the other two profiles, Dual Messenger apps appear in most list of apps you might come across in system settings, mixed in with your main user apps, rather than separated. As mentioned in one of the issues linked above, some activity launcher apps (but not all) end up launching the main user's equivalent activity instead (for some but not all activities), despite using a copy of the activity launcher installed on the Dual Messenger profile—something the copies of the apps in the other two profiles can never do.

In the case of opening and sharing apps like OpenLinkWith and URL Checker, I've found that the list of available applications they display when running the copy of the app installed in the Dual Messenger profile is that of the main user, rather than correctly showing the apps installed to the Dual Messenger profile. This makes it impossible or extremely difficult to send some links to the correct app in the Dual Messenger profile.

I realize this is a very circumstantial issue, and one that depends on using an official One UI feature in a very non-approved way, but it would be quite useful to the few of us who use these programs this way too figure out exactly what makes the Dual Messenger profile so different, and how to get it working as it should.

Let me know if I can clarify anything further here. 

### Expected behavior

URL Checker, when installed to User 95 (configured via Samsung's Dual Messenger) shows a list of other apps installed to User 95 when using the "Open with" feature (Share uses the system share menu for User 0, but this doesn't matter as User 95's apps are included in this list, as mentioned above).

### Actual behavior

URL Checker instead only shows a list of User 0's apps, despite being installed to User 95.

### URLCheck version

2.9.1

### Android version

13

### Android Custom/Specific ROM or Device

One UI

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I will fill out all of the requested information in this form.

**Comment by TrianguloY** [comment]
Incredible work both on the research and the explanation! 
I checked the linked issues, and it seems it's a common issue, although apparently you can "fix" it sometimes by installing the app as dual itself, or having a proxy to redirect the opening versions.

I'm very interested in a solution, not for this app (well, that too) but for another one of mine, because being able to detect and open a dual version is like the holy grail of Samsung. 

I'm not sure if I can be on any help, if there is a way I can try to implement it, and if there is something I can test just explain and I'll do my best. Not sure about how the proxy activity should work, is there any app out there with a solution?  That ActivityLauncher seems like it, but with all the ghost messages it was a bit hard to follow the conversation :/

**Comment by TrianguloY** [comment]
I didn't know 1 was already "solved", I'll try to check how those other apps are able to detect dual versions (thanks for the links!)

As for 2, my two cents: apparently when you use Samsung's official share dialog, the dual versions are present because Samsung itself injects them. This seems to be a Samsung specific "feature" not available to other apps. 
There is a strange package called "samsung da agent" that apparently works as an official proxy for dual apps (it even had a security issue in the past if I'm not mistaken). I haven't been able to "use" it though, but maybe it is possible.
